### ISSUE TITLE
Exchange mechanism principles

### ISSUE THREAD

**Issue by kvoer** [stichtingsem/ecosystem#7]
### What are the underlying principles in the exchange of data between Roles and their specific APIs?

This issue continues the discussion as mentioned in issue [7](https://github.com/stichtingsem/technology-prototype/issues/7) from the PoC phase 1 project.

Questions to be answered:
- What mechanism do we choose to exchange data in which process?
- Does a sales order require a push kind of exchange because of legal constraints?
- How do we cope with initial setup of new parties joining the ecosystem?
- What are the expected volumes of each data exchange within the ecosystem? Does this change our view of the preferred mechanism?
- What data do we exchange in the event info? Are we using ReferenceID or is the actual data included?

We should get a clear overview of the following table:

| Process| Sender | Receiver | API | Content | Mechanism | Event content | Volume |
|---|---|---|---|---|---|---|---|
| All | SIS | MP, LA, LMS | SIS Data | All | Webhook and events | ReferenceID | ? |
| New order | MP | LA | Entitlement | Sales orders | ? | ? | ? |
| All | MP | LMS | Entitlement | All | Webhook and events | ReferenceID | ? |
| All | LA | MP, LMS | Catalogue | All | Webhook and events | ReferenceID | ? |
| All | LA | MP, LMS | Usage| All | Webhook and events | ReferenceID | ? |
| All | LA | MP, LMS | Progress | All | Webhook and events | ReferenceID | ? |
| All | LA | MP, LMS | Results | All | Webhook and events | ReferenceID | ? |

**Comment by dpronk** [comment]
Also on a generic level, how do we arrange / force things like schema validation ? are we going to add json schema's to the stack ?

**Comment by kvoer** [comment]
During the Technical Track meeting @edwinverwoerd suggested to add a calculator to the documentation where new parties can see how much load they can expect in a specific exchange.

**Comment by TM-Frank** [comment]
Discussion (1):
Event API :”Order should always be reverse chronological.”
I suggest that events will be followed by an API call (mostly), and in that way always the latest data is retrieved from te source. So chronological events are not needed and wil be an extra complexity.

Discussion (2):
Event API: “Providers should ensure that this API is cacheable or able to be retrieved at minimal cost, as it may be accessed a lot by consumers who are 'catching up'
If the providers is responsable for facilitating an “event-log”, why don't subscribers retrieve directly from this eventlog, (authorised on events) in stead of point to point event communication tot subscribers?

**Comment by edwinverwoerd** [comment]
All Entitlements will have the Mechanism of POST/PULL.

**Comment by edwinverwoerd** [comment]
There should be a  mechanism that a service can collect all the (last) available data. Same status in time and data if a new service will do there bootstrap.

**Comment by edwinverwoerd** [comment]
Please change proces New Order to all! Because order api is not yet a part of the ECO system. (strange because within the current ECK set-up we have one?)

**Comment by mcginkel** [comment]
I dont understand the new order thing. this need clarification

Events are always in the order that they are created. So you get the events in creation order. You cannot get update events, after delete events. sequence is important.
If the data is send in the event, you can process immediatelly.
If you missed events and need to catchup: catchup and process events
If you try to get the object via a get method then you get the latest anyway (with all updates).

**Comment by mcginkel** [comment]
"All Entitlements will have the Mechanism of POST/PULL."

What do you mean here? My understanding is that the entitlements are that important (cost/good/money) that you want the MP to deliver them to the LA. The event mechanism will do exactly that: it will push the changes to the entitlement as events.
If you change the spec if the eventhandler from the LA that the events for entiotlements MUST be processed synchroiously or fail you have the garanteed delivery.

PULL is still dependent on the LA pulling the entitlements. POST is basicly the same as events, but with another name/mechanisme.

**Comment by mcginkel** [comment]
"There should be a mechanism that a service can collect all the (last) available data. Same status in time and data if a new service will do there bootstrap."

You can use the registration of an eventhook or the consent API to trigger the flow of events (sending all) so the parties are up to date.
Example: school admin gives consent to the LA to use student data. When the consent is successfull, the SIS can send all events ones to this LA.

**Comment by mcginkel** [comment]
"Also on a generic level, how do we arrange / force things like schema validation ? are we going to add json schema's to the stack ?"
The event has an event type. That will determine the schema of the "data" tag. We can document that explicitly

**Comment by mcginkel** [comment]
proposal:


| Process| Sender | Receiver | API | Content | Mechanism | Event content | Volume |
|---|---|---|---|---|---|---|---|
| All | SIS | MP, LA, LMS | SIS Data | All | Webhook and events | actual data in event| 1 per student, 1 per group, 1 per teacher, foolowe dby updates |
| LML / New orders | MP | LA | Entitlement | Sales orders | forced events (sync) | actual data in event | ? |
| All | MP | LMS | Entitlement | All | Webhook and events | actual data in event | ? |
| All | LA | MP, LMS | Catalogue | All | Webhook and events | actual data in event| ? |
| All | LA | MP, LMS | Usage| All | Webhook and events | actual data in event | ? |
| All | LA | MP, LMS | Progress | All | Webhook and events | actual data in event| ? |
| All | LA | MP, LMS | Results | All | Webhook and events | ?| ? |

**Comment by hhaarsma** [comment]
From LA perspective the events for Progress and Results worry me. We will be communicating about groups of assignments/interactions, and the progress and results will be on that group level. But everytime one of the assignments/interactions will be finished this will result in a progress update, so an event.
When we talk about some 100.000 sessions per day, that will generate quite a lot of events, and will also place quite some requirements on storing all of those events.
So I am wondering/thinking how we can prevent that we create a system that will be unusable because of the high numbers.

Can we use some mechanism to limit the number of events that we will be generating?
How long should we store the events?

**Comment by edwinverwoerd** [comment]
Maybe back to the old school. Let de progress app do a give me the progress or result for this person. I do not think that I by myself will use grouplevel.

**Comment by edwinverwoerd** [comment]
Transaction from an entitlement will always be a POST too a LMS LA SA.

**Comment by kvoer** [comment]
We addressed this issue in the standard version 1.1. The architecture for all APIs in the Ecosystem is Event Driven Architecture. This also applies to the Entitlement API. We added EntitlementConfirmation events to this API to meet the compliance and audit needs of the MP.

In the specification we added two temporary POST endpoints for Entitlements and Confirmations. These Endpoints are only used within POC2 and will be deprecated in Q3 of 2022.
